File: pkg/functionRuntimeInterface/functionRuntimeInterface.go
package functionRuntimeInterface

import (
	"bufio"
	"context"
	"fmt"
	"github.com/rs/zerolog/log"
	"os"
	"strings"
	"time"

	pb "github.com/3s-rg-codes/HyperFaaS/proto/function"
	zerolog "github.com/rs/zerolog"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

type handler func(context.Context, *Request) (*Response, error)

type Request struct {
	Data string
	Id   string
}

type Response struct {
	Data string
	Id   string
}

type Function struct {
	timeout  int
	address  string
	request  *Request
	response *Response
	id       string
}

func New(timeout int) *Function {
	address, ok := os.LookupEnv("CALLER_SERVER_ADDRESS")
	if !ok {
		log.Error().Msgf("Environment variable CALLER_SERVER_ADDRESS not found")
	}

	return &Function{
		timeout:  timeout,
		address:  fmt.Sprint(address, ":50052"),
		request:  &Request{},
		response: &Response{},
		id:       getID(),
	}
}

func (f *Function) Ready(handler handler) {

	//Set up logging
	log := configLog("logs", "fn.log")

	connection, err := grpc.NewClient(f.address, grpc.WithTransportCredentials(insecure.NewCredentials()))

	log.Debug().Msgf("Connected to the worker: %v", f.address)

	defer connection.Close()

	if err != nil {
		log.Error().Msgf("failed to connect: %v", err)
	}

	c := pb.NewFunctionServiceClient(connection)

	//Set the id in the response to the id of the container
	f.response.Id = f.id

	defer log.Info().Msgf("Closing connection.")
	first := true

	for {
		ctx, cancel := context.WithTimeout(context.Background(), time.Duration(f.timeout)*time.Second)
		defer cancel()

		//We ask for a new request whilst sending the response of the previous one
		p, err := c.Ready(ctx, &pb.Payload{Data: f.response.Data, Id: f.response.Id, FirstExecution: first})
		first = false

		if err != nil {
			log.Error().Msgf("failed to call: %v", err)
			return
		}
		log.Debug().Msgf("Received request: %v", p.Data)

		f.request = &Request{p.Data, p.Id}

		f.response, err = handler(ctx, f.request)

		log.Debug().Msgf("Function handler called and generated response: %v", f.response.Data)

		if err != nil {
			log.Error().Msgf("Function failed: %v", err)
			return
		}

	}

}

func configLog(logDir string, logFile string) *zerolog.Logger {
	// Open the log file
	file, err := os.OpenFile(logDir+"/"+logFile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)

	if err != nil {
		panic(err)
	}

	log := zerolog.New(file).With().Timestamp().Logger()
	return &log
}

func getID() string {
	var id string
	file, err := os.Open(".env")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasPrefix(line, "CONTAINER_ID=") {
			id = strings.TrimPrefix(line, "CONTAINER_ID=")
			break
		}
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}
	return id
}

File: functions/python/functionRuntimeInterface.py
import contextvars
import logging
import os
from dataclasses import dataclass
from typing import Callable

import grpc

from proto.function import function_pb2_grpc, function_pb2


@dataclass
class Response:
    data: str
    id: str


@dataclass
class Request:
    data: str
    id: str


Handler = Callable[[contextvars.ContextVar, Request], Response]


def function_runtime_interface(handler: Handler):
    address = os.getenv("CALLER_SERVER_ADDRESS")
    if address is None:
        address = ""
        address = "{}:50052".format(address)
    ide = get_id()

    with (grpc.insecure_channel(address) as channel):
        c = function_pb2_grpc.FunctionServiceStub(channel)

        ctx = contextvars.ContextVar('func')
        firstexec = True
        while True:
            if firstexec:
                p = c.Ready()
            logging.debug(f"Received request: {p.data}")
            resp = handler(ctx, Request(id=p.id, data=p.data))
            logging.debug(f"Function handler called and generated response: {resp.data}")
            payload = function_pb2.Payload(data=resp.data, id=resp.id, firstExecution=firstexec)
            firstexec = False
            p = c.Ready(payload)


def get_id():
    id = ""
    try:
        with open('.env', 'r') as file:
            for line in file:
                if line.startswith("CONTAINER_ID="):
                    id = line.strip().split("CONTAINER_ID=")[1]
                    break
    except Exception as e:
        raise e

    return id
